File: TerminBA/TerminBA.Services/Database/FacilityDynamicPrice.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Linq.Expressions;
using System.Xml;

namespace TerminBA.Services.Database
{
    public class FacilityDynamicPrice
    {
        [Key]
        public int Id { get; set; }

        [Required]
        [ForeignKey(nameof(Facility))]
        public int FacilityId { get; set; }
        public Facility? Facility { get; set; }

        [Required]
        public DayOfWeek StartDay { get; set; }
        [Required]
        public DayOfWeek EndDay { get; set; }

        [Required]
        public TimeOnly StartTime { get; set; }
        [Required]
        public TimeOnly EndTime { get; set; }

        [Required]
        [Column(TypeName = "decimal(10,2)")]
        public decimal PricePerHour { get; set; }

        [NotMapped]
        public bool IsActive =>
              (ValidFrom <= DateOnly.FromDateTime(DateTime.Today)) &&
              (ValidTo == null || ValidTo >= DateOnly.FromDateTime(DateTime.Today));

        public static Expression<Func<FacilityDynamicPrice, bool>> IsActiveExpr(DateOnly today) =>
            x => x.ValidFrom <= today &&
            (x.ValidTo == null || x.ValidTo >= today);

        [Required]
        public DateOnly ValidFrom { get; set; }

        public DateOnly? ValidTo { get; set; }
    }
}

File: TerminBA/TerminBA.UI/terminba_sport_center_desktop/lib/model/facility_dynamic_price.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:terminba_sport_center_desktop/enums/day_of_week_enum.dart';

part 'facility_dynamic_price.g.dart';

@JsonSerializable()
class FacilityDynamicPrice {
	int id;
	int facilityId;
	String? facilityName;
	@JsonKey(fromJson: dayOfWeekFromJson, toJson: dayOfWeekToJson)
	DayOfWeek startDay;
	@JsonKey(fromJson: dayOfWeekFromJson, toJson: dayOfWeekToJson)
	DayOfWeek endDay;
	String startTime;
	String endTime;
	double pricePerHour;
	bool isActive;
	DateTime validFrom;
	DateTime? validTo;

	FacilityDynamicPrice({
		required this.id,
		required this.facilityId,
		this.facilityName,
		required this.startDay,
		required this.endDay,
		required this.startTime,
		required this.endTime,
		required this.pricePerHour,
		required this.isActive,
		required this.validFrom,
		this.validTo,
	});

	factory FacilityDynamicPrice.fromJson(Map<String, dynamic> json) =>
			_$FacilityDynamicPriceFromJson(json);

	Map<String, dynamic> toJson() => _$FacilityDynamicPriceToJson(this);
}
